### PR TITLE
build: Improvements to Codespaces

### PR DESCRIPTION
**Pull request by alexec**
Fixes #TODO

Please do not open a pull request until you have checked ALL of these:

* [ ] Create the PR as draft .
* [ ] Run `make pre-commit -B` to fix codegen and lint problems.
* [ ] Sign-off your commits (otherwise the DCO check will fail).
* [ ] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [ ] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [ ] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [ ] Github checks are green.
* [ ] Once required tests have passed, mark your PR "Ready for review".

If changes were requested, and you've made them, dismiss the review to get it reviewed again.

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
-    "onCreateCommand": ".devcontainer/pre-build.sh",
-    "workspaceMount": "source=${localWorkspaceFolder},target=${localEnv:HOME}/go/src/github.com/argoproj/argo-workflows,type=bind",
-    "workspaceFolder": "${localEnv:HOME}/go/src/github.com/argoproj/argo-workflows"
+  "image": "mcr.microsoft.com/vscode/devcontainers/base:ubuntu",
+  "features": {
+    "ghcr.io/devcontainers/features/go:1": {
+      "version": "1.19"
+    },
+    "ghcr.io/devcontainers/features/node:1": {
+      "version": "18"
```
Comment (alexec) [inline .devcontainer/devcontainer.json:8]: Change node to v18 which is (a) LTS (b) does not have the hash function issue.

File: .dockerignore [hunk 2]
```diff
@@ -3,6 +3,8 @@
 *.iml
 *.md
 *.yaml
+.devcontainer
+.git
```
Comment (alexec) [inline .dockerignore:7]: Excluding `git`  is a bigger change than it looks. The `.git` folder gets modified every time your run `git` preventing caching of builds from working. Once removed, we must pass `GIT_COMMIT` as Docker build-args.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -1,7 +1,18 @@
 {
-    "build": { "dockerfile": "Dockerfile" },
-    "hostRequirements": {"cpus": 4},
-    "onCreateCommand": ".devcontainer/pre-build.sh",
-    "workspaceMount": "source=${localWorkspaceFolder},target=${localEnv:HOME}/go/src/github.com/argoproj/argo-workflows,type=bind",
-    "workspaceFolder": "${localEnv:HOME}/go/src/github.com/argoproj/argo-workflows"
+  "image": "mcr.microsoft.com/vscode/devcontainers/base:ubuntu",
```
Comment (alexec) [inline .devcontainer/devcontainer.json:2]: The previous "universal" image does not work on Apple Silicon.

File: tasks.yaml [hunk 4]
```diff
@@ -18,26 +18,19 @@ spec:
   tasks:
     - name: go-deps
       command: go mod download
-      mutex: deps
-    - name: static-files
-      command: make server/static/files.go STATIC_FILES=false
-      dependencies: go-deps
-    - command: go build -v ./...
```
Comment (alexec) [inline tasks.yaml:25]: this is actually quite slow, so I removed it

File: ui/package.json [hunk 5]
```diff
@@ -6,13 +6,13 @@
         "src"
     ],
     "scripts": {
-        "build": "rm -Rf dist && NODE_OPTIONS='--openssl-legacy-provider' NODE_ENV=production webpack --mode production --config ./src/app/webpack.config.js",
```
Comment (alexec) [inline ui/package.json:9]: Fixed by reverting to Node 16.

**Comment by agilgur5** [comment]
Replaced by #10599

**Pull request closed without merging**